nltk/metrics/distance.py

```diff
@@ -182,19 +187,125 @@ def custom_distance(file):
             labelA, labelB, dist = l.strip().split("\t")
             labelA = frozenset([labelA])
             labelB = frozenset([labelB])
-            data[frozenset([labelA,labelB])] = float(dist)
-    return lambda x,y:data[frozenset([x,y])]
+            data[frozenset([labelA, labelB])] = float(dist)
+    return lambda x, y: data[frozenset([x, y])]
 
 
-def demo():
-    edit_distance_examples = [
-        ("rain", "shine"), ("abcdef", "acbdef"), ("language", "lnaguaeg"),
-        ("language", "lnaugage"), ("language", "lngauage")]
-    for s1, s2 in edit_distance_examples:
-        print("Edit distance between '%s' and '%s':" % (s1, s2), edit_distance(s1, s2))
-    for s1, s2 in edit_distance_examples:
-        print("Edit distance with transpositions between '%s' and '%s':" % (s1, s2), edit_distance(s1, s2, transpositions=True))
+def jaro_similarity(s1, s2):
+    """
+    Computes the Jaro similarity between 2 sequences from:
+
+        Matthew A. Jaro (1989). Advances in record linkage methodology
+        as applied to the 1985 census of Tampa Florida. Journal of the
+        American Statistical Association. 84 (406): 414-20.
+
+    The Jaro distance between is the min no. of single-character transpositions
+    required to change one word into another. The Jaro similarity formula from
+    https://en.wikipedia.org/wiki/Jaro%E2%80%93Winkler_distance :
+
+        jaro_sim = 0 if m = 0 else 1/3 * (m/|s_1| + m/s_2 + (m-t)/m)
+
+    where:
+        - |s_i| is the length of string s_i
+        - m is the no. of matching characters
+        - t is the half no. of possible transpositions.
+    """
+    # First, store the length of the strings
+    # because they will be re-used several times.
+    len_s1, len_s2 = len(s1), len(s2)
+
+    # The upper bound of the distance for being a matched character.
+    match_bound = max(len_s1, len_s2) // 2 - 1
+
+    # Initialize the counts for matches and transpositions.
+    matches = 0  # no.of matched characters in s1 and s2
+    transpositions = 0  # no. of transpositions between s1 and s2
+    flagged_1 = []  # positions in s1 which are matches to some character in s2
+    flagged_2 = []  # positions in s2 which are matches to some character in s1
+
+    # Iterate through sequences, check for matches and compute transpositions.
+    for i in range(len_s1):     # Iterate through each character.
+        upperbound = min(i+match_bound, len_s2-1)
+        lowerbound = max(0, i-match_bound)
+        for j in range(lowerbound, upperbound+1):
+            if s1[i] == s2[j] and j not in flagged_2:
+                matches += 1
+                flagged_1.append(i)
+                flagged_2.append(j)
+                break
+    flagged_2.sort()
+    for i, j in zip(flagged_1, flagged_2):
+        if s1[i] != s2[j]:
+            transpositions += 1
```

Comment:
This loop isn't correct, you're checking non-consistent order of list of positions. E.g.

```
s1_matching_position = [0, 1, 2, 3, 5]
s2_matching_position = [1, 2, 3, 8]
```

This will have zero matches although 1, 2, 3 is the match.

Reply:
@alvations  , [this paper by WW Cohen : ](www.cs.cmu.edu/~wcohen/postscript/ijcai-ws-2003.pdf) suggests to arrange the matching positions of s1 and s2 in the same order as they have appeared and then to compare the corresponding characters at those positions.

Reply:
k = trans_count = 0
    for i, ying_f in enumerate(ying_flags):
        if ying_f:
            for j in _range(k, yang_len):
                if yang_flags[j]:
                    k = j + 1
                    break
            if ying[i] != yang[j]:
                trans_count += 1
    trans_count /= 2


@alvations ,This code snippet is from another [implementation of Jaro_Winkler](https://github.com/jamesturk/jellyfish/blob/master/jellyfish/_jellyfish.py) . 

Here **ying** and **yang** are same as **s1** and **s2** respectively.
 Also **ying_flags** and **yang_flags** are same as **flagged_s1** and **flagged_s2** respectively. 

The implementation is exactly same with this PR.

Reply:
@alvations , any updates?

Reply:
@alvations ,[this paper by WW Cohen](www.cs.cmu.edu/~wcohen/postscript/ijcai-ws-2003.pdf) suggests  that the matches in both the strings must be arranged in the order in which they appear in each of the strings. Now , `flagged_1` is already in the sorted order since `s1` is traversed according to its order of occurrence. Also doing `flagged_2.sort()` actually sorts the positions of matches of `s2 `in the order of occurrence.

Also this is a code snippet from [another implementation of Jaro winkler distance]( https://github.com/jamesturk/jellyfish/blob/master/jellyfish/_jellyfish.py). 
    
```
    k=trans_count=0
    for i, ying_f in enumerate(ying_flags):
        if ying_f:
            for j in _range(k, yang_len):
                if yang_flags[j]:
                    k = j + 1
                    break
            if ying[i] != yang[j]:
                trans_count += 1
    trans_count /= 2
```
    
Here `ying `and `yang` are same as` s1` and `s2 `respectively. Also `ying_flags` and `yang_flags `are same as` flagged_s1` and `flagged_s2`. 

This implementation is exactly same as this PR

Reply:
Ah got it. The `s1[i] != s2[j]` is adding the transposition counts. Sorry about it. Didn't read it carefully.